demo.html

```diff
@@ -0,0 +1,262 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <meta charset="utf-8">
+        <title>Demo Grapher</title>
+    </head>
+    <body>
+        <p>
+            Enter a JS math equation<br>using x, y, k, and theta:
+        </p>
+        <div style="color: white; z-index: 5;">
+            <input style="border-radius: 1rem;" id="ieq" value="Math.sin(k*x)*Math.cos(y/k)" type="text">
+            <br>
+            <button style="border-radius: 1rem;" onclick="handleInput();" type="button" id="1">Save</button>
+            <button style="border-radius: 1rem; display:none;" onclick="cancel(); handleInput();" type="button" id="2">Save</button>
+        </div>
+        <canvas style="width: 80%; position: absolute; top: 50%; left: 50%; transform: translate(-50%, -50%);" id="bg">Start streaming on a modern browser!</canvas>
+    </body>
+    <script>
+        var equation
+        const handleInput = () => {
+            savebutton = document.getElementById("1")
+            savebutton.style.display = "none"
+            againbutton = document.getElementById('2')
+            againbutton.style.display = "block"
+            equation = document.getElementById("ieq").value
+            animateGraph()
+        }
+
+        const animateGraph = () => {
+            var el = document.getElementById("bg")
+            // Make sure to keep this width-height 3:2 ratio
+            el.width = 2100
+            el.height = 1400
+
+            const c = el.getContext("2d")
+            const w = c.canvas.width
+            const h = c.canvas.height
+
+            c.scale(1, -1)
+            c.translate(0, -h)
+           
+            // x-range of graph
+            const x_Ra = [-5, 5]
+            // y-range of graph
+            const y_Ra = [-5, 5]
+
+            var lightSource = [2.5, 2.5, 2.5]
+            var i_hat = [1, 0]
+            var j_hat = [1255.77, 0.999]
+
+            var k = 1
+            var f = (x, y) => k * eval(equation)
+            var v = true
+
+            var theta, tanTheta, tanThetaSquared
+            theta = tanTheta = tanThetaSquared = 0
+
+            var blend = (c, int) => {
+
+                var mi = 0
+                var max = 0
+                for (var i = 0; i < c.length; i++) {
+
+                    if (c[i] > max) {
+
+                        mi = i
+                        max = c[i]
+
+                    }
+
+                }
+                var int255 = 255 * int
+                return [int255 * c[0] / c[mi], int255 * c[1] / c[mi], int255 * c[2] / c[mi]]
+
+            }
+
+            var brightness = (x, y, z, lTheta, lPhi) => {
```

Comment:
Just a suggestion, but for the optimizations, removing some of the tedious calculations in these shading functions and replacing them with a simple function would only minimally affect the shading but lead to drastic speed increases.

Reply:
Yup! I've added a high/low quality option

Reply:
Looks good